File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.cpp
typedef long long int lli;
const lli inf = 1e6;

class Solution {
public:
	long long countSubarrays(vector<int>& nums, int minK, int maxK) {
		lli ret = 0;
		int n = nums.size();

		vector<vector<lli>> v;
		vector<lli> temp;
		for (int i = 0; i < n; i++) {
			if (nums[i] < minK || nums[i] > maxK) {
				if (!temp.empty()) v.push_back(temp);
				temp.clear();
			} else {
				temp.push_back(nums[i]);
			}
		}
        if (!temp.empty()) v.push_back(temp);
		for (auto& it : v) {
			vector<lli> max_idx, min_idx;
            
			for (int i = 0; i < it.size(); i++) {
                // cout << i << endl;
				if (it[i] == maxK) {
					max_idx.push_back(i);
				}

				if (it[i] == minK) {
					min_idx.push_back(i);
				}
				if (!max_idx.empty() && !min_idx.empty()) {
					lli idx = min(max_idx.back(), min_idx.back());
                    // cout << idx << " -> "<< max_idx.back() << " " << min_idx.back() << endl;
					ret += idx + 1;
                    
				}
			}
            // cout << endl;
		}
		return ret;
	}
};

File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.java
class Solution {
    public long countSubarrays(int[] nums, int minK, int maxK) {
        int left = -1, min = -1, max = -1, n = nums.length;
        long count = 0l;
        for (int i = 0; i < n; i++) 
            if (nums[i] < minK || nums[i] > maxK) 
                left = min = max = i;
            else {
                min = nums[i] == minK ? i : min;
                max = nums[i] == maxK ? i : max;
                count += (min < max ? min : max) - left;
            } 
        return count;
    }
}